### ISSUE TITLE
Structure of Manifolds, Restructure Packages.

### ISSUE THREAD

**Issue by kellertuer** [JuliaManifolds/Manifolds.jl#119]
We are currently again running into the timing issue in testing and I feel like it's time to split a little further. In general I think it's worth to go to a 3-level-approach

## `ManifoldsBase.jl`

This basic interface to host all basis functions and definitions. This should include the basic functions so for example also `hat` and `vee` (which are currently in `Manifolds.jl`s base file even) as well as basic notions, for example I would like to also move types like `QRRetraction` over there since it is in some sense a basic ingredient type used by manifolds and maybe  

## `ManifoldsDecorator.jl`

This one can maybe also called `ManioldsMeta.jl` since it's in general more about adding things to manifolds or deriving things from it. This should just be using `ManifoldsBase.jl` and contain (in my oppinion)

* power and product manifold
* the metric manifolds interface
* the embedded manifolds interface
* the group manifolds interface (though not the specific groups)

and is hence some kind of a meta middle level. If you just think ybout basic manifolds, the first should be fine

## `Manifolds.jl`

Maybe this now comes closer to what we originally thought of when naming it `ManifoldsMuseum`, this uses the above two to implement concrete manifolds and to host the documentation for all three.

## `Manopt.jl`
A fourth package that fits this regime is my optimization toolbox, which definetly fits into the regime of applying the first two/three packages

## Machine Learning on Manifolds
This is first mentioned in #73 though I am not sure about the current status, I would love to see this in action :)

## Distributions on Manifolds
I think also the distributions are independent of the library.

## Conclusion
I am still very thankful to AntoineLevit for getting us started here at JuliaNLSolvers. We don't really fit into the solvers regime, but for start it was awesome. With these many packages in this plan, maybe it's time to start a `JuliaManifolds` or `JuliaNLSpaces` Team?

Of course this is all not in a hurry and we should definetly wait until we have a first version out before moving – or is it better to move (just) before we register a new package? The rework to three packages can be postponed to after a first release. Moving would just mean restarting a register and maybe pushing a minor version.

**Comment by sethaxen** [comment]
I approve of this and I agree with the proposed schedule. I moved a registered package to another org, and it was easy.

**Comment by kellertuer** [comment]
So do you think we should move before the first public version? Might be better to get also the org name out and have a fixed url. I think with the embedding now finished (just merging needed) and basis closing in, we could announce a first registered version.

**Comment by sethaxen** [comment]
If we're decided that we will move to an org sometime soon, we might as well do it first for simplicity. I've never set up an org, but I bet it's pretty straightforward to do and migrate.

As for org name, JuliaManifolds sounds good to me. What are the pros/cons of that vs JuliaNLSpaces?

**Comment by kellertuer** [comment]
My favourite would be JuliaManifolds, though it doubles in name with the main package then., but I think that's also no too bad.

**Comment by mateuszbaran** [comment]
I also like  JuliaManifolds as the organization name :+1: . We could make it even now and transfer the repositories once the two big PRs are merged (to avoid any disruptions in CI).

I can also add `FunManifolds.jl` to it once it's updated (though I'd like to finish moving my non-functional stuff to Manifolds.jl before it happens).

Let's now go through our dependencies:
```
Arpack = "7d9fca2a-8960-54d3-9f78-7d1dccf2cb97"
```
Probably a direct dependency just to fix it's version. Indirectly required by `Distributions`.
```
Distributions = "31c24e10-a181-5473-b8eb-7969acd0382f"
Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
SpecialFunctions = "276daf66-3868-5448-9aa4-cd146d93841b"
StatsBase = "2913bbd2-ae8a-5f71-8c99-4fb6c76f3a91"
```
Required by distributions and statistics-related functionality. We could probably move them to a separate package that builds upon `Manifolds.jl` itself.
```
Documenter = "e30172f5-a6a5-5a46-863b-614d45cd2de4"
```
Probably not necessary now since we have it as a dependency in docs `Project.toml`?
```
Einsum = "b7d42ee7-0b51-5a75-98ca-779d3107e4c0"
```
Required by the metric manifold.
```
FiniteDifferences = "26cc04aa-876d-5657-8c51-4c34ba976000"
Requires = "ae029012-a4dd-5104-9daa-d747884805df"
```
Required by differentiation stuff. It should be relatively high in our hierarchy of packages, though preferably not in Base. Definitely `ManifoldDecorators.jl` though if it happens.
```
FunctionWrappers = "069b7b12-0de2-55c6-9aab-29f3d0a68a2e"
```
Probably a requirement just to fix its version due to being required by a package used in the test suite?
```
StaticArrays = "90137ffa-7385-5640-81b9-e52037218182"
HybridArrays = "1baab800-613f-4b0a-84e4-9cd3431bfbb9"
```
Essentially required by the power manifold.
```
LightGraphs = "093fc24a-ae57-5d10-9952-331d41423f4d"
SimpleWeightedGraphs = "47aef6b3-ad0c-573a-a1e2-d07658019622"
```
Required by the graph manifold
```
LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
ManifoldsBase = "3362f125-f0bb-47a3-aa74-596ffd7ef2fb"
Markdown = "d6f4376e-aef5-505a-96c1-9c027394607a"
Statistics = "10745b16-79ce-11e8-11f9-7d13ad32a3b2"
```
Very basic.
```
UnsafeArrays = "c4a57d5a-5b31-53a6-b365-19f8c011fbd6"
```
May not be needed any more, I just didn't take the time to make sure.

I'd say from looking at this statistics and distribution-related things could be separated. I'm not sure about the benefits of splitting decorators. This would make any changes to the API harder to coordinate. Do you want to split it just to reduce the CI times?

**Comment by kellertuer** [comment]
Cool that you are directly also thinking about dependencies! Reorganizing is a good idea to also reduce these and get us up to maybe not having Manifest.toml Yes we could also do that once basis is done, so before the first release.

Concerning the split: Yes it originated in trying to get CI times down and make the project a little more modular. Yes, changing things in Base might affect updating Meta and Manifolds then, that's right. I am open to any other suggestion to reduce CI times for the main package though.

Conceptually you're right, 3 levels is not so nice, we could also add the explicit decorators (not only the interface) to Base, i.e. EmbeddedManifold, MetricManifold and GroupManifold, but that might be too much in Base then? I am not yet sure there.

**Comment by mateuszbaran** [comment]
I'll have to check at one point what exactly is taking so much time during testing. Probably mostly compilation times.

EmbeddedManifold and GroupManifold could be to some degree moved to Base but MetricManifold has a few additional dependencies so I wouldn't add it to Base.

**Comment by kellertuer** [comment]
So we have the new org, yay! And I am not that sure anymore about the 3-level structure, maybe that might cause too much trouble when updating the lowest (Base) level package. But I think we can put some decorators in separate independent files, for example the group manifold, maybe even together with all groups (so its based on `ManifoldsBase` and `Manifolds`)?

Further, I think my main point of this issue is already resolved with the new org.

**Comment by mateuszbaran** [comment]
With the recently reduced testing time there is now less pressure to split the package. Having it in one repository makes making bigger changes more simple. Anyway I think we should eventually move a couple of things away from `Manifolds.jl` but it's definitely not urgent. I don't know if it's a good enough reason to keep this issue open or not.

**Comment by sethaxen** [comment]
> With the recently reduced testing time there is now less pressure to split the package

I haven't been following the recent PRs. What caused the reduced testing time?

**Comment by mateuszbaran** [comment]
Not testing by default on `MArray` and floats other than `Float64` together with adding `optimize=0` Julia flag to our CI scripts.

**Comment by kellertuer** [comment]
...You can still activate all these tests by setting the corresponding flags in `runtests.jl`, though. So you can run them locally and we could activate them on the CI again, whenever we think its fast enough or the package is small enough.

**Comment by kellertuer** [comment]
I think this is resolved, since we have the own org, we will stick to the two-level approach (ManifoldsBase and packages building on that).
